### PR TITLE
Added tests for encoding and decoding functions

### PR DESCRIPTION
**Pull request by multivac61**

### CODE REVIEW

**Comment by ibab** [comment]
I'm currently writing some tests for encoding/decoding myself :)
But I've just chained encode->decode, which should complement your tests.

File: test/test_decoding.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+            # Encode and decode audio input 
+            encoded_audio_manual = manual_encode(audio, mu)
+            decoded_audio_manual = manual_decode(encoded_audio_manual, mu)
+
+            decoded_audio_net_tensor = self.net.decode(encoded_audio_manual)
+            decoded_audio_net = decoded_audio_net_tensor.eval()
+
+            # Shapes should be the same
+            self.assertAllEqual(decoded_audio_net.shape, decoded_audio_manual.shape)
```
Comment (ibab) [inline test/test_decoding.py]: There's no need to test the shapes, as `assertAllClose` does this automatically.

**Comment by multivac61** [comment]
Ah ok great!

I find it very weird that the Travis CI build is failing. Commited tests all run without errors on my machine, both using `nodetest` and `nodetest3`. I am running a virtualenv with tensorflow and all pip packages within [requirements.txt](https://github.com/ibab/tensorflow-wavenet/blob/master/requirements.txt) installed.

File: test/test_decoding.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    return x
+
+
+class TestDecode(tf.test.TestCase):
+
+    def setUp(self):
+        quantization_steps=256
+        self.net = WaveNet(batch_size = 1,
+                           channels = quantization_steps,
+                           dilations= [1, 2, 4, 8, 16, 32, 64, 128, 256,
```
Comment (ibab) [inline test/test_decoding.py]: As we don't really need the network itself, you could set `dilations=[1]` or maybe even `dilations=[]` to save some time in building it.

File: test/test_decoding.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    x = np.sign(y) * (np.exp(y * np.log(1 + mu)) - 1) / mu
+
+    return x
+
+
+class TestDecode(tf.test.TestCase):
+
+    def setUp(self):
+        quantization_steps=256
+        self.net = WaveNet(batch_size = 1,
```
Comment (ibab) [inline test/test_decoding.py]: The spacing around the `=` is inconsistent here. I would personally prefer no spaces around `=` for kwargs, like in PEP8.

**Comment by ibab** [comment]
@multivac61: Yeah, please rename `_preprocess` to `encode` and move both to `wavenet_ops.py` :)
I'll have a look at whether I can figure out why it's failing on travis.

File: test/test_decoding.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+from wavenet import WaveNet
+from test_encoding import manual_encode
+import json
+import numpy as np
+
+def manual_decode(signal, channels):
+    # Calculate inverse mu-law companding and dequantization
+    mu = channels
+    output = signal.astype(np.float32)
+    y = (2*output - 1) / mu
```
Comment (ibab) [inline test/test_decoding.py]: Shouldn't this be `2 * (output / mu) - 1`?
It was recently fixed in this commit: https://github.com/ibab/tensorflow-wavenet/commit/9f85246631d9ffc3d76ae881442c993a5f6e91b6
I think the reason travis is failing is that there's a mismatch between the manual decoding and what's now on `master`.

**Comment by multivac61** [comment]
Just [commited code ](https://github.com/ibab/tensorflow-wavenet/pull/52/commits/6404a79c467437f132e6b3462a7aa24fc5695dac) edited based on your comments. Should pass the travis-CI tests after the bug on [decode](https://github.com/ibab/tensorflow-wavenet/blob/master/wavenet.py#L81) is fixed.

Another PR request for code refactoring is underway!

**Comment by ibab** [comment]
I think the bug you are referring to has already been fixed on `master`, so you just need to rebase to get the latest version.

**Comment by ibab** [comment]
I've pushed my work on a different branch here: https://github.com/ibab/tensorflow-wavenet/commit/42991d00c1c493c7ca4d173d5f4a58cffd4e7a20
I'll try to merge these changes after this PR.

**Comment by multivac61** [comment]
I have made a [commit](https://github.com/multivac61/tensorflow-wavenet/commit/6404a79c467437f132e6b3462a7aa24fc5695dac) where I addressed the changes you requested.

**Comment by ibab** [comment]
@multivac61: Yeah, I think rebasing on `master` is the best option at this point.

**Comment by ibab** [comment]
It seems that you've collected quite a few commits from the master branch.
Not sure what will happen if we merge this :/

**Comment by multivac61** [comment]
Ah I see. I must admit that I am a total noob when it comes to git. But I am learning and find this project really interesting and want to contribute to it, I am not  giving up.

**Comment by ibab** [comment]
This can be closed now, as we've merged #84.

**Pull request closed without merging**